run-tests.sh

```diff
@@ -13,17 +13,33 @@ unset MAKELEVEL
 
 # Get the directory of the current script, in a POSIX compatible way
 # https://stackoverflow.com/a/29835459
-script_directory=$(CDPATH= cd -- "$(dirname -- "$0")" && pwd -P)
+script_directory=$(CDPATH="$(cd -- "$(dirname -- "$0")")" && pwd -P)
 
 # Set this variable to ensure top level Makefile doesn't affect the results.
 export PROJECT_ROOT="${script_directory}/tests"
 
 describe() {
-  printf "\n$BOLD$1$RESET"
+  printf "\n%s%s%s" "$BOLD" "$1" "$RESET"
 }
 
 it() {
-  printf "\n  $1"
+  printf "\n  %s" "$1"
+}
+
+trim_string() {
```

Comment:
Why is this `trim_string` function necessary?

Reply:
I'm not 100% sure why it happened (my suspicion is something weird with POSIX shell stuff, I tested it with dash) but all tests weren't passing because for some reason every RUN_RESULT had an extra space at the end... I didn't touch the functions that set that variable so POSIX weirdness is my theory. I guess it's also smart to trim the results for tests to make them more strict so that's another reason to do it like this.

I could look into it more, but from what I've looked through before, I couldn't come up with a good answer as to why it kept adding a space at the end.

Reply:
Does it also happen when you run the tests on the `main` branch with `dash`? The CI [runs the tests with `dash`](https://github.com/paldepind/projectdo/blob/18d950c3a26ee1f951fed52caaaf20c5712bce89/.github/workflows/makefile.yml#L15) and I've run the tests with several shells (including dash) on both Linux and macOS without problems.

Reply:
Huh weird, it passes them just fine on main. I wonder what causes that extra space... I'll try to figure it out but the only thing I changed was that I added quotes around $1 in the subshells: `RUN_RESULT=$(cd tests/"$1" && ../../projectdo -n build)` That shouldn't effect it though right?

Reply:
Yeah, that sounds a bit weird. But in that case I think it would be best to figure out what caused the extra space and then not have this trim function (seems simpler overall) :)

Reply:
This is a really weird problem. I agree that it's simpler without it so I would like to fix this too, but I don't know what's causing this. I ruled out the $1 -> "$1" change, it's not causing this. For some reason `$(cd tests/"$1" && ../../projectdo -n run)` and others are all appending a single space to the end. I can't replicate it in an interactive shell (dash), just the script does this.

Reply:
I always heard of echo having undefined behavior, but I never experienced it till now. In `execute()` function there was an echo statement where I quoted the `$TOOL_ARGS` variable. Apparently echo was printing a space even when `$TOOL_ARGS` was empty when quoted. I now replaced it with printf. I needed to also add an if statement, because without it it would concatenate the `$1` and `$TOOL_ARGS` and fail the tests. Everything works now and I removed the `trim_string()` function.

Reply:
I think that is expected actually. `echo` prints a space between each argument. If you do `echo "foo" "bar"` it will print `foo bar` so it makes sense that if you do `echo "foo" ""` then it will print `echo foo `. The last argument happens to be an empty string, but it is still given two arguments so a space is printed in between. With the old syntax we where actually passing `echo` a variable number of arguments and then it worked out.

In any case, great to avoid the trim function 👍